Fix major version check for stateless environment

vapopov opened the pull request:
In this PR local process storage replaced with the backend storage with proper resource serialization. For current self-hosted clusters preserved last known version must be migrated to backend storage.

Related: https://github.com/gravitational/teleport/issues/49848
Changelog: Fixed major version check for stateless environment

vapopov committed:
Added auth information resource with persisting teleport version

vapopov commented:
We would probably need to check inventory store to identify version sets for the auth servers, for instance if we have in cluster `auth1:v17.0.0`, `auth2:v18.0.0` - auth1 should be restricted to be downgraded to v16, and auth2 to be upgraded to v18. Not clear if we able to retrieve full information from inventory stream in auth initialization step. tbd

vapopov commented:
Also was trying in parallel PR to reuse current `PresenceService` which stores the Teleport version in `auth_server` spec (to not create additional resource), but `LocalSupervisor` keep updating this info at every start even if auth service is failed to init.

vapopov committed:
Check the set of the auth info to compare with min/max versions in cluster

hugoShaka reviewed (changes requested):
If I understand correctly, during the startup:
- the auth infos are listed on startup to discover the lowest and highest major versions
- teleport writes its own authinfo resource with a TTL of 24h

I see several issues with the current approach:
- the auth info write happens on startup, if no auth got restarted in the last 24 hours, all authinfo resources are expired and the check becomes useless as it cannot figure the previous auth version
- it is not possible to perform sequentially two major updates the same day

Note: there's a variant of the first issue where no auth were running in the last 24 hours, this often happens when people forget to clear their dynamo backend and a new cluster reuses the old backend. So making the authInfo writes periodic might not completely solve the issue.

I don't have another solution to suggest yet but I would expect the version check to have the following properties:
- consistently works even if the cluster was downscale for a week
- supports every officially supported version transitions (e.g. updating from v15 to v16 and v16 to v17 the same day, as long as no 2 auths are running concurrently)

vapopov commented:
@hugoShaka I was mostly worried case when after downscale you wont be able to update other auth instances because of existing records of downscaled instances, this is why expiry is set. We might just make it simple as possible and always check per host id instead of range of versions (but this don't prevent the case when we have v17 and v16 instances running in cluster to downgrade v16 one to v15, there might be the issues with roles registration in storage for instance, like deprecating or adding new role in new major version).

hugoShaka commented:
> We might just make it simple as possible and always check per host id instead of range of versions

This will not work on stateless deployments as hostIDs are not persisted.

hugoShaka commented:
I wonder if a single non-expiring cluster-wide BackendInfo resource would work. Something like:
- auth starts
- auth gets BackendInfo
  - if BackendInfo version is too low we crash
  - if BackendInfo is lower than us, we update backend info
  - if BackendInfo is higher than us, we don't change anything
- auth continues to start and runs its migrations
- (optionally) auth updates BackendInfo again to note that it finished applied its migrations

With optimistic locking we should be protected against races. The resource would represent pretty closely the backend state.

Another variant would be to create one BackendInfo per Teleport version, to list all backend infos and to look at what's the highest. This would be closer to a regular database migration mechanism and the resource would represent the history of the backend migrations, which might be useful.

cc @espadolini @rosstimothy and @fspmarshall

espadolini commented:
You can't do conditional operations on key ranges (and even fetching key ranges is sort of wobbly and slightly inconsistent, time-wise), so the logic becomes a lot more awkward if we have to make decisions based on a range of items.

I'm +1 for recording "the version" of the cluster in a single item and conditionally updating it at auth start, with the understanding that this only acts as a lint and not as a hard compatibility guarantee (the only way to truly do that would be to make every read and write an atomic operation that also checks the version item, AFAICT). We could reuse `--skip-version-check` (which only currently works on proxies and peripheral agents) to bypass the check to allow for downgrading, too.

vapopov committed:
Store only one entity for the cluster version

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/major-version-check-v2

hugoShaka commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+	if err := authinfotype.ValidateAuthInfo(info); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	version, err := semver.NewVersion(info.GetSpec().TeleportVersion)
```
Nit: we can also use `version.EnsureSemver()` which is more tolerant regarding the `v` prefix.

hugoShaka commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+	return version, nil
+}
+
+// WriteTeleportVersion writes the last known Teleport version to the storage.
+func (s *AuthInfoService) WriteTeleportVersion(ctx context.Context, version *semver.Version) error {
+	if version == nil {
```
I think we should stick with RFD 156 and expose the full Create Update Read Delete operations. Each time we've done differently we ended up needs every operation at some point.

espadolini replied:
I don't know if it's necessary but it wouldn't hurt.

hugoShaka commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+		Expires:  expires,
+		Revision: rev,
+	}
+	if _, err := s.backed.Put(ctx, item); err != nil {
+		return trace.Wrap(err)
+	}
```
The unconditional put makes racing possible, someone might have updated the version in the meantime. This operation should use conditional updates based on revision. I think we have all the machinery available in the generic service struct.

vapopov replied:
I'm Changing this one, but not sure if it is good to fail auth init because when we do parallel scale of auth instances and spawning same auth versions in cluster, they have to conditionally write to the same key and theoretically if they launched and the same second one of the instances must fail to write and restart

hugoShaka reviewed (commented):
This looks way better, thanks for taking the time to do all the changes again.

cc @espadolini for another review 🙏

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// WriteTeleportVersion writes the last known Teleport version to the storage.
+func (s *AuthInfoService) WriteTeleportVersion(ctx context.Context, version *semver.Version) error {
```
Just take a `semver.Version` by value? It's just a handful of scalar fields.

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+
+// AuthInfoService is responsible for managing the information about auth server.
+type AuthInfoService struct {
+	backed backend.Backend
```
```suggestion
	backend backend.Backend
```

espadolini commented on api/types/authinfo/info.go:
```diff
@@ -0,0 +1,57 @@
+// ValidateAuthInfo checks that required parameters are set
+// for the specified AuthInfo.
+func ValidateAuthInfo(c *authinfo.AuthInfo) error {
+	if c == nil {
+		return trace.BadParameter("AuthInfo is nil")
+	}
+	if c.Metadata == nil {
+		return trace.BadParameter("Metadata is nil")
+	}
+	if c.Spec == nil {
+		return trace.BadParameter("Spec is nil")
+	}
```
None of these conditions are dealbreakers for protobuf messages, and we should be using getters specifically to avoid having to check for `nil` everywhere.

espadolini commented on lib/auth/version.go:
```diff
@@ -36,19 +38,32 @@ const (
+	lastKnownVersion, err := backendStorage.GetTeleportVersion(ctx)
 	if trace.IsNotFound(err) {
-		if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
+		// TODO(vapopov): DELETE IN v19.0.0 last known version must be already migrated to backed storage.
```
```suggestion
		// TODO(vapopov): DELETE IN v19.0.0 last known version must be already migrated to backend storage.
```

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+)
+
+const (
+	// authInfoPrefix is a backend key for storing the auth server information.
+	authInfoPrefix = "auth_info"
```
If it's the whole key it's not a prefix.

espadolini replied:
How about something like this?

```suggestion
	// authInfoKeyComponent is the name of the backend item for storing the auth server information.
	authInfoKeyComponent = "auth_info"
```

espadolini commented on api/types/authinfo/info.go:
```diff
@@ -0,0 +1,57 @@
+	if c.Spec == nil {
+		return trace.BadParameter("Spec is nil")
+	}
+
```
We're checking some structural `nil` pointers but we're not checking the actually important things (i.e. kind, subkind and version).

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+	if err := authinfotype.ValidateAuthInfo(info); err != nil {
+		return trace.Wrap(err)
+	}
+	rev, err := types.GetRevision(info)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
The revision is (rightfully) ignored on writes other than `ConditionalUpdate`, so this is unnecessary; also, for a known concrete type you can just do `info.GetMetadata().GetRevision()`.

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,111 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	expires, err := types.GetExpiry(info)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
I would just not support expiry for the auth_info singleton.

espadolini commented on api/proto/teleport/authinfo/v1/authinfo.proto:
```diff
@@ -0,0 +1,37 @@
+
+// AuthInfoSpec encodes the parameters auth server meta-information.
+message AuthInfoSpec {
+  // teleport_version advertises the version of the auth server.
```
I would clarify what sort of string we expect, if it's a semver string starting with `v` or not.

vapopov committed:
Add CRUD endpoints
Change to use conditional update and retry
Fix spelling

vapopov committed:
Add validation for version, sub kind, kind, name

vapopov committed:
Rename to authinfo.go

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/major-version-check-v2

vapopov committed:
Assigning error after creating new resource

vapopov commented:
@espadolini @hugoShaka could you please have another look?

espadolini commented on api/proto/teleport/authinfo/v1/authinfo.proto:
```diff
@@ -0,0 +1,37 @@
+
+// AuthInfoSpec encodes the parameters auth server meta-information.
+message AuthInfoSpec {
+  // teleport_version advertises the version of the auth server, e.g. "17.3.3".
```
I would consider stressing that this does not have a leading "v", i.e. it's a string that's valid according to coreos/go-semver.

espadolini commented on api/types/authinfo/info.go:
```diff
@@ -0,0 +1,65 @@
+// for the specified AuthInfo.
+func ValidateAuthInfo(info *authinfo.AuthInfo) error {
+	switch {
+	case info == nil:
+		return trace.BadParameter("AuthInfo is nil")
+	case info.GetMetadata() == nil:
+		return trace.BadParameter("Metadata is nil")
+	case info.GetSpec() == nil:
+		return trace.BadParameter("Spec is nil")
```
```suggestion
```

espadolini commented on api/types/authinfo/info.go:
```diff
@@ -0,0 +1,65 @@
+	case info.GetSpec() == nil:
+		return trace.BadParameter("Spec is nil")
+	case info.GetKind() != types.KindAuthInfo:
+		return trace.BadParameter("wrong AuthInfo Kind: %s", info.Kind)
```
Use `%q` (or `%+q` for ascii-only, even) to format strings that we're not sure of.

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,151 @@
+		return semver.Version{}, trace.Wrap(err)
+	}
+
+	tVersion, err := version.EnsureSemver(info.GetSpec().TeleportVersion)
```
If we've defined `teleport_version` to be a semver-valid string we don't need to check and handle a missing leading "v" here.

espadolini replied:
```suggestion
	tVersion, err := version.EnsureSemver(info.GetSpec().GetTeleportVersion())
```

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,151 @@
+)
+
+const (
+	// maxWriteRetry is the number of retries for conditional updates of the AuthInfo resource.
+	maxWriteRetry = 5
```
This is scoped to the whole `lib/services/local` package which is unfortunately pretty huge, give it a more specific name.

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,151 @@
+// WriteTeleportVersion writes the last known Teleport version to the backend.
+func (s *AuthInfoService) WriteTeleportVersion(ctx context.Context, version semver.Version) (err error) {
+	var info *authinfo.AuthInfo
+	for i := 0; i < maxWriteRetry; i++ {
```
```suggestion
	for range maxWriteRetry {
```

espadolini commented on lib/services/local/authinfo.go:
```diff
@@ -0,0 +1,151 @@
+	return *tVersion, nil
+}
+
+// WriteTeleportVersion writes the last known Teleport version to the backend.
+func (s *AuthInfoService) WriteTeleportVersion(ctx context.Context, version semver.Version) (err error) {
```
If this is just going to write the version you can just use upsert, the point of conditional operations is that you read the current value, you make a decision based on the value and then you update the value as long as nothing else changed it. With this implementation, a cluster running 16.3.0 that happens to launch a 16.4.2 and a 17.3.3 auth at the same time might end up with a stored version of 16.4.2 if the write of the first auth ends up happening later than the write of the second.

This function should be called `UpdateTeleportVersion` or something along those lines, and it should take the revision of the previous auth_info resource.

vapopov replied:
make sense, I've moved retry logic to version check function, where we request auth_info resource with revision and retry if its already created or already updated by another process

vapopov committed:
Move retry logic to version check helper

espadolini commented on lib/auth/version.go:
```diff
@@ -20,55 +20,101 @@ package auth
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
```
What's the point of creating the `auth_info` with the stored version just to immediately write it again?

I would consider moving the migration part of this out of the loop and maybe have it work in the opposite way: if there's a version in process storage then create it in the backend (ignoring AlreadyExists) and then delete the version from process storage, then proceed with the regular logic.

vapopov replied:
if there many auth servers we are going to continuously override the version in backend until every instance drop their local last known version

espadolini commented on lib/auth/version.go:
```diff
@@ -20,55 +20,101 @@ package auth
+					err = trace.Wrap(err)
+					slog.WarnContext(ctx, "Failed to create AuthInfo resource",
+						"error", err)
+					continue
```
We should only retry concurrency errors here, not everything (i.e. `AlreadyExists` from create and `CompareFailed` - or even `backend.ErrIncorrectRevision` - from update).

espadolini commented on lib/auth/version.go:
```diff
@@ -20,55 +20,101 @@ package auth
+	backendStorage services.AuthInfoService,
+	currentVersion semver.Version,
+) (err error) {
 	if skip := os.Getenv(skipVersionUpgradeCheckEnv); skip != "" {
 		return nil
 	}
```
Should we forcibly update the value in the backend if the envvar is set? There's no good way to "heal" the value if there's some problem, otherwise.

espadolini commented on lib/auth/version.go:
```diff
@@ -20,55 +20,101 @@ package auth
+				lastKnownVersion, currentVersion.String(), lastKnownVersion.Major-1)
+		}
+
+		authInfo.GetSpec().TeleportVersion = currentVersion.String()
+		if _, err := backendStorage.UpdateAuthInfo(ctx, authInfo); err != nil {
+			err = trace.Wrap(err)
+			slog.WarnContext(ctx, "Failed to update AuthInfo resource",
+				"error", err)
+			continue
+		}
```
We only need to write the version if it's different (higher?), this write can be skipped if the version is the same.

espadolini commented on lib/auth/version.go:
```diff
@@ -20,55 +20,101 @@ package auth
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
+				lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
+		}
+		if lastKnownVersion.Major-currentVersion.Major > 1 {
```
```suggestion
		if lastKnownVersion.Major > currentVersion.Major {
```

vapopov replied:
why? we should restrict only if MajorVersion is higher or lower than 1. 15 -> 16 (allow), 15 -> 17 (deny)

espadolini replied:
This is the downgrade protection, not the upgrade version change limit. We should disallow downgrades across major versions without acknowledging the risk.

vapopov replied:
got you, previously we want to allow major version downgrade

espadolini replied:
Downgrades are not really supported in general, downgrades past a major version even less so. The escape hatch of `--skip-version-check` is always there for those scenarios, right?

vapopov replied:
changed: https://github.com/gravitational/teleport/pull/52837/commits/9b40b98eb237467500cd378ddecc68d2c605a6ba

espadolini commented on lib/auth/version.go:
```diff
@@ -20,55 +20,101 @@ package auth
-		return trace.Wrap(err)
+		lastKnownVersion, err := semver.NewVersion(authInfo.GetSpec().GetTeleportVersion())
+		if err != nil {
+			return trace.Wrap(err)
```
This error condition deserves a better error message and some steps that the cluster admin can take to fix the situation (broken backend item or invalid version string).

vapopov committed:
Call create/update depends on if resource is created already
Read local database once without retry

vapopov committed:
Restrict major version downgrade

vapopov committed:
Make `--skip-version-check` available for major upgrade check

vapopov committed:
Fix linter warnings

vapopov commented:
@hugoShaka @espadolini Could you please take another look when you have a moment?

hugoShaka reviewed (approved):
Looks good, I have to other optional remarks:

- It feels like AuthInfo is now more like BackendInfo as it's a singleton
- It might be useful to have a way to check the authinfo with tctl, either in the status command, or the resource one. Because this PR is already quite large and we dragged the review process, it would be nice to do this is a followup PR

hugoShaka commented on lib/auth/version.go:
```diff
@@ -20,55 +20,100 @@ package auth
+				return trace.Wrap(err)
+			}
+		}
+		return nil
```
We raced 😅  https://github.com/gravitational/teleport/pull/52837#discussion_r2064437498

vapopov commented on lib/auth/version.go:
```diff
@@ -20,55 +20,100 @@ package auth
-	}
-	if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
-		return trace.Wrap(err)
+	var authInfo *authinfov1.AuthInfo
+	for range versionUpgradeCheckMaxWriteRetry {
+		authInfo, err = backendStorage.GetAuthInfo(ctx)
```
we have to declare `authInfo` because errors from `GetAuthInfo`, `CreateAuthInfo`, `UpdateAuthInfo` must be returned after last retry, otherwise it going to be shadowed

vapopov commented on lib/auth/version.go:
```diff
@@ -20,55 +20,100 @@ package auth
+		return nil
 	}
-	return nil
+	return
```
here we created/updated authInfo with current version and we need to exit from retry loop

vapopov committed:
Add logs and make skip more safe in case of broke item in backend

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/major-version-check-v2

vapopov committed:
Remove deleting version item from process database, stateBackend doesn't support deletion (requires implementation for kube secrets storage)

vapopov committed:
Rename AuthInfo to BackendInfo
Add cleanup of the version item from process storage after successful migration

vapopov committed:
Make skip-version-check to upsert the version in backend

vapopov commented:
@espadolini @hugoShaka hope I've addressed all your comments, if I miss something let me know

espadolini commented on lib/auth/version.go:
```diff
@@ -20,55 +20,136 @@ package auth
+
+		// TODO(vapopov): DELETE IN v19.0.0 – the last known version should already be migrated to backend storage.
+		if cleanProcVersion {
+			if err := procStorage.DeleteTeleportVersion(ctx); err != nil {
+				return trace.Wrap(err)
+			}
```
extreme nit just in case we have an early restart of the auth and two concurrent processes on the same process storage:
```suggestion
			if err := procStorage.DeleteTeleportVersion(ctx); err != nil && !trace.IsNotFound(err) {
				return trace.Wrap(err)
			}
```

vapopov committed:
Update lib/auth/version.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>
